### PR TITLE
Add x:SetterTargetType

### LINKED ISSUE #6960: Allow Setters outside of a Style
E. g.
```xaml
<Button>
  <Interaction.Behaviors>
     <local:StartAnimationOnClick>
       <Animation Duration="0:0:1"> 
         <KeyFrame Cue="0%">
           <Setter Property="Opacity" Value="1.0"/>
         </KeyFrame>
         <KeyFrame Cue="50%">
           <Setter Property="Opacity" Value="0.5"/>
         </KeyFrame>
         <KeyFrame Cue="100%">
           <Setter Property="Opacity" Value="1.0"/>
          </KeyFrame>
        </Animation>
     </local:StartAnimationOnClick>
  </Interaction.Behaviors>
<Button>
```

We probably need some `x:SetterTargetType` hint attribute to support cases like this.

### PR DESCRIPTION
**Pull request by pr8x**
Adding `x:SetterTargetType` to allow for using `<Setter>` outside of `<Style>`.

Issues:

- > ERROR: XAMLIL Unable to find emitter for node type: XamlX.Ast.XamlAstXmlDirective Line 6, position 6. in D:\Repositories\Avalonia\src\Avalonia.Build.Tasks\bin\Debug\net6.0\MainWindow.axaml 6:6-6:6

  Other transformers seem to remove the `XamlAstXmlDirective` from the children collection. I tried doing that as well, but for some reason it processes the node twice and on the second run it will fail to find the attribute and fail the compilation. 

   Repro:  [av_repro_SetterTargetType_emit_error.patch.txt](https://github.com/AvaloniaUI/Avalonia/files/9167344/av_repro_SetterTargetType_emit_error.patch.txt)

- It seems that VS does not pick up this new attribute. How do I register it properly?

  image

CC: @grokys @kekekeks 

Fix #6960

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0022325-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by maxkatz6** [comment]
> It seems that VS does not pick up this new attribute. How do I register it properly?

Looks like a Resharper highlighting. Am I right? In this case JetBrains would need to support it.

File: src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlSetterTransformer.cs [hunk 1]
```diff
@@ -22,21 +19,54 @@ public IXamlAstNode Transform(AstTransformationContext context, IXamlAstNode nod
...
-                .FirstOrDefault(p => p.Property.GetClrProperty().Name == "Selector");
-            if (selectorProperty == null)
-                throw new XamlParseException(
-                    "Can not find parent Style Selector", node);
-            var selector = selectorProperty.Values.FirstOrDefault() as XamlIlSelectorNode;
-            if (selector?.TargetType == null)
-                throw new XamlParseException(
-                    "Can not resolve parent Style Selector type", node);
+
+            if (styleParent != null)
```
Comment (maxkatz6) [inline src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlSetterTransformer.cs:28]: There are conflicts after ControlTheme PR.

**Comment by pr8x** [comment]
> Looks like a Resharper highlighting. Am I right

True, seems to be some ReSharper weirdness.

File: src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlSetterTransformer.cs (hunk 1 above)
Comment (pr8x) [reply src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlSetterTransformer.cs:28]: Conflicts should be fixed now

**Comment by jmacato** [comment]
@pr8x just curious on what's the supposed use case for this one?

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0022466-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by pr8x** [comment]
@jmacato Constructing  `<Animation>` outside of  `<Styles>` (e.g. inside Resources or for assigning to property)

**Comment by jmacato** [comment]
that's a good use case! thanks for the example!

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0022486-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by pr8x** [comment]
Removing the `XamlAstXmlDirective` works in the UT, but still fails when I try to do it in Sandbox (see repro patch in description). For some reason the transformer is called twice...

**Comment by maxkatz6** [comment]
@pr8x test fails if you copypaste this sandbox patch to the unit test. Except you need to remove "x:Class" part. Test also should be initialized with `using (UnitTestApplication.Start(TestServices.StyledWindow))`.

From what I checked, if fails when there are two or more setters in the animation. Sandbox code with single keyframe works fine.

**Comment by maxkatz6** [comment]
@pr8x I think probably better way would be to one more transformer as well similar to AvaloniaXamlIlControlThemeTransformer, and replace Animation with AvaloniaXamlIlTargetTypeMetadataNode by filling ScopeTypes with a new enum value and TargetType.
Compiler should process this node normally then, and it will be able to read TargetType from there for multiple setters.
At least this way ControlTheme and Style fethes

**Comment by pr8x** [comment]
@maxkatz6 I wonder why it tries to emit something for these directive nodes at all? Couldn't it just ignore them? Wouldn't we have the same issue for `AvaloniaXamlIlTargetTypeMetadataNode` that it tries to emit something for these nodes?

**Comment by maxkatz6** [comment]
> What if in the future we want to add a transformer for Animation?

From what I understand, AvaloniaXamlIlTargetTypeMetadataNode special node with metadata just wraps original one. And original one is still available and can be transformed later. @kekekeks if you can confirm it.

**Comment by maxkatz6** [comment]
@pr8x @kekekeks I added a new transformer to inject AvaloniaXamlIlTargetTypeMetadataNode with the parsed type. Please take a look.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0025280-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0025413-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by lindexi** [comment]
> > Looks like a Resharper highlighting. Am I right
> 
> True, seems to be some ReSharper weirdness.

Reference: https://youtrack.jetbrains.com/issue/RSRP-498037/Cannot-resolve-xSetterTargetType-in-Avalonia-XAML

**Pull request merged**